mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -506,37 +506,104 @@ def _get_deleted_runs(self):
-            # we assume that the ``latest_metrics`` table already accounts for its presence
-            if just_created:
-                self._update_latest_metric_if_necessary(logged_metric, session)
+            try:
+                run = self._get_run(run_uuid=run_id, session=session)
+                self._check_run_is_active(run)
+
+                # metrics are grouped by metric name for last metric calculation
+                metrics_per_key = {}
+
+                for metric in metrics:
+                    _validate_metric(metric.key, metric.value, metric.timestamp, metric.step)
+                    is_nan = math.isnan(metric.value)
+                    if is_nan:
+                        value = 0
+                    elif math.isinf(metric.value):
+                        #  NB: Sql can not represent Infs = > We replace +/- Inf
+                        #  with max/min 64b float value
+                        value = (
+                            1.7976931348623157e308 if metric.value > 0 else -1.7976931348623157e308
+                        )
+                    else:
+                        value = metric.value
+
+                    # ToDo: Consider prior checks for null, type, metric name validations, ... etc.
+
+                    if batch_mode:
+                        # In batch mode, all metrics are added to the session.
+                        # This is much faster than checking for each metric if a metric with the
+                        # same name, value, ts, and step already exists and adding them otherwise.
+                        # Conflicts will be checked during saving. If conflicts are detection
+                        # during saving, metrics are inserted one after another
+                        # using _get_or_create
+
+                        logged_metric = SqlMetric(
+                            run_uuid=run_id,
+                            key=metric.key,
+                            value=value,
+                            timestamp=metric.timestamp,
+                            step=metric.step,
+                            is_nan=is_nan,
+                        )
+                        self._save_to_db(session, logged_metric)
+                        just_created = True
+
+                    else:
+                        # All metrics are added one after another with additional check if a
+                        # metric with same name, value, ts, and step already exists
+
+                        logged_metric, just_created = self._get_or_create(
+                            model=SqlMetric,
+                            session=session,
+                            run_uuid=run_id,
+                            key=metric.key,
+                            value=value,
+                            timestamp=metric.timestamp,
+                            step=metric.step,
+                            is_nan=is_nan,
+                        )
```

Comment:
Unfortunately, `_get_or_create` is much slower because an additional query operation is performed for every metric (checking if it already exists). The same check is performed however when saving a session (due to the constraint check).

Using only `_get_or_create` takes about 2 seconds for 2000 points while inserting in this way takes only 0.2 seconds (if no conflicts are in the data which I guess is the case most of the time).